src/TestFramework/Coverage/CoveredFileData.php

```diff
@@ -0,0 +1,138 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Coverage;
+
+use Infection\TestFramework\Coverage\XmlReport\FileCodeCoverage;
+use Symfony\Component\Finder\SplFileInfo;
+use Webmozart\Assert\Assert;
+
+/**
+ * Workhorse AKA envelope for all things coverage-related in regard of mutators.
+ *
+ * @internal
+ * @final
+ */
+class CoveredFileData implements LineCodeCoverage
+{
+    /**
+     * @var SplFileInfo
+     */
+    private $sourceFile;
+
+    /**
+     * @var CoverageFileData|null
+     */
+    private $coverageFileData;
+
+    /**
+     * @var iterable<CoverageFileData>
+     */
+    private $lazyCoverageFileData;
+
+    /**
+     * @var FileCodeCoverage|null
+     */
+    private $lineCodeCoverage;
+
+    /**
+     * @param iterable<CoverageFileData> $lazyCoverageFileData
+     */
+    public function __construct(SplFileInfo $sourceFile, iterable $lazyCoverageFileData)
+    {
+        $this->sourceFile = $sourceFile;
+
+        // There's no point to have it parsed right away as we may not need it, e.g. because of a filter
+        $this->lazyCoverageFileData = $lazyCoverageFileData;
+    }
+
+    public function getSplFileInfo(): SplFileInfo
+    {
+        return $this->sourceFile;
+    }
+
+    public function getRealPath(): string
+    {
+        $realPath = $this->sourceFile->getRealPath();
+
+        Assert::string($realPath);
+
+        return $realPath;
+    }
+
+    /**
+     * Accessor used to update CoverageFileData with TestFileTimeData.
+     */
+    public function retrieveCoverageFileData(): CoverageFileData
```

Comment:
I'm not quite sure about that `FileCodeCoverage <-> CoveredFileData` relationship/API, will get back to it later